### PR TITLE
tests/fuzzers/les: add fuzzer for les server handler

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR refactors the LES server handler in order to make fuzzer testing possible and implements the fuzzer.
Replaces https://github.com/ethereum/go-ethereum/pull/22173 and https://github.com/ethereum/go-ethereum/pull/22237

### CODE REVIEW

File: les/server_handler.go [hunk 1]
```diff
@@ -223,648 +221,109 @@ func (h *serverHandler) handleMsg(p *clientPeer, wg *sync.WaitGroup) error {
 	}
 	defer msg.Discard()
 
-	var (
-		maxCost uint64
-		task    *servingTask
-	)
 	p.responseCount++
```
Comment (rjl493456442) [inline les/server_handler.go:224]: Is it problematic? The description to `responseLock` is 

```
// responseLock ensures that responses are queued in the same order as
// RequestProcessed is called
```

while the responseCount here is bumped without lock protection.

File: les/server_handler.go [hunk 2]
```diff
@@ -223,648 +221,109 @@ func (h *serverHandler) handleMsg(p *clientPeer, wg *sync.WaitGroup) error {
...
 	responseCount := p.responseCount
-	// accept returns an indicator whether the request can be served.
-	// If so, deduct the max cost from the flow control buffer.
-	accept := func(reqID, reqCnt, maxCnt uint64) bool {
-		// Short circuit if the peer is already frozen or the request is invalid.
-		inSizeCost := h.server.costTracker.realCost(0, msg.Size, 0)
```
Comment (rjl493456442) [inline les/server_handler.go:236]: Would be super super nice if we can encapsulate these "before-serving" logic and "after-serving" logic into two separate functions: "prepare"  and "finalize". The execution is in the middle.

Just throw out the thoughts here. We can open the following PR if you also think it's a good idea.

**Comment by rjl493456442** [comment]
Ran the fuzzer for a while, no crash detected. 

```
2021/02/19 15:05:51 workers: 8, corpus: 20 (10m36s ago), crashers: 0, restarts: 1/8950, execs: 948729 (783/sec), cover: 3369, uptime: 20m12
```

File: les/server_handler.go (hunk 1 above)
Comment (zsfelfoldi) [reply les/server_handler.go:224]: `handleMsg` is called sequentially so it does not need to be thread safe and this is the only place where `p.responseCount` is accessed. The lock is used in the serving goroutine and what it really ensures is that `RequestProcessed` and `queueSend` calls are done in the same order. So I think this one is ok (it works the same way in the original version).

File: les/server_handler.go (hunk 2 above)
Comment (zsfelfoldi) [reply les/server_handler.go:236]: Sure, let's do it in a follow-up PR.

File: les/server_handler.go (hunk 1 above)
Comment (rjl493456442) [reply les/server_handler.go:224]: Right. Resolved it.

**Review by rjl493456442 — APPROVED** [review]
LGTM

**Pull request merged**